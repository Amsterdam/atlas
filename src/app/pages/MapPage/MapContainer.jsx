import React from 'react'
import MapPanel from '../../../map/containers/panel/MapPanel'
import {
  getMapBaseLayers as fetchBaseLayers,
  getPanelLayers as fetchPanelLayers,
  getMapLayers as fetchMapLayers,
} from '../../../map/services'
import { encodeLayers, decodeLayers } from '../../../store/queryParameters'
import getState from '../../../shared/services/redux/get-state'
import MapPage from './MapPage'
import MapContext from './MapContext'

function setQueryStringParameter(name, value) {
  const params = new URLSearchParams(window.location.search)

  params.set(name, value)

  window.history.pushState({}, '', decodeURIComponent(`${window.location.pathname}?${params}`))
}

export function getQueryStringParameter(name) {
  const params = new URLSearchParams(window.location.search)
  return params.get(name)
}

export function getActiveBaseLayer() {
  return getQueryStringParameter('achtergrond') || 'topografie'
}

export const DEFAULT_LAT = 52.3731081
export const DEFAULT_LNG = 4.8932945

const initialState = {
  activeBaseLayer: 'topografie',
  activeMapLayers: [],
  baseLayers: [],
  viewCenter: [DEFAULT_LAT, DEFAULT_LNG],
  zoomLevel: 11,
}

const reducer = (state, action) => {
  switch (action.type) {
    case 'setActiveBaseLayer':
      return {
        ...state,
        activeBaseLayer: action.payload,
      }
    case 'getBaseLayers':
      return {
        ...state,
        baseLayers: action.payload,
      }
    case 'getPanelLayers':
      return {
        ...state,
        panelLayers: action.payload,
      }
    case 'getMapLayers':
      return {
        ...state,
        mapLayers: action.payload,
      }
    case 'setActiveMapLayers':
      // TODO: simplify this in the new MapPanel
      return {
        ...state,
        activeMapLayers: state.activeMapLayers.some((overlay) =>
          action.payload.map(({ id }) => id).includes(overlay.id),
        )
          ? [
              ...state.activeMapLayers.filter((overlay) => {
                return !action.payload.map(({ id }) => id).includes(overlay.id)
              }),
            ]
          : [
              ...state.activeMapLayers,
              ...action.payload
                .reverse()
                .map(({ id: mapLayerId, isVisible }) => ({ id: mapLayerId, isVisible })),
            ],
      }
    case 'setVisibleMapLayers':
      // TODO: simplify this in the new MapPanel
      return {
        ...state,
        activeMapLayers: state.activeMapLayers.map((overlay) => ({
          ...overlay,
          isVisible:
            overlay.id === action.payload.mapLayerId ? action.payload.isVisible : overlay.isVisible,
        })),
      }
    case 'setMapPanelVisible':
      return {
        ...state,
        isMapPanelVisible: action.payload,
      }
    default:
      return state
  }
}

const MapContextProvider = ({ children }) => {
  const [state, dispatch] = React.useReducer(reducer, initialState)

  // Get the user from the redux state
  const { user } = getState()

  function setActiveBaseLayer(payload) {
    if (typeof window !== 'undefined') {
      setQueryStringParameter('achtergrond', payload)
    }

    dispatch({ type: 'setActiveBaseLayer', payload })
  }

  function setActiveMapLayers(payload) {
    dispatch({ type: 'setActiveMapLayers', payload })
  }

  function setVisibleMapLayers(mapLayerId, isVisible) {
    dispatch({ type: 'setVisibleMapLayers', payload: { mapLayerId, isVisible: !isVisible } })
  }

  function setMapPanelVisible() {
    if (typeof window !== 'undefined') {
      setQueryStringParameter('legenda', !state.isMapPanelVisible)
    }

    dispatch({ type: 'setMapPanelVisible', payload: !state.isMapPanelVisible })
  }

  async function getBaseLayers() {
    const baseLayers = await fetchBaseLayers()

    dispatch({
      type: 'getBaseLayers',
      payload: baseLayers.reduce(
        (result, item) => ({
          ...result,
          [item.category]: result[item.category] ? [...result[item.category], item] : [item],
        }),
        {},
      ),
    })
  }

  async function getPanelLayers() {
    const panelLayers = await fetchPanelLayers()

    dispatch({
      type: 'getPanelLayers',
      payload: panelLayers,
    })
  }

  async function getMapLayers() {
    const mapLayers = await fetchPanelLayers()

    dispatch({
      type: 'getMapLayers',
      payload: mapLayers,
    })
  }

  // Update the state from the url with the maplayers
  React.useEffect(() => {
    if (typeof window !== 'undefined') {
      const activeMapLayers = getQueryStringParameter('lagen')

      setActiveMapLayers(decodeLayers(activeMapLayers))
    }
  }, [window.location.search])

  // Update the url with the maplayers
  React.useEffect(() => {
    if (typeof window !== 'undefined') {
      // if (state.activeMapLayers.length > 0)
      //   setQueryStringParameter('lagen', encodeLayers(state.activeMapLayers))
    }
  }, [state.activeMapLayers])

  return (
    <MapContext.Provider
      value={{
        ...state,
        user,
        setActiveBaseLayer,
        setActiveMapLayers,
        setVisibleMapLayers,
        setMapPanelVisible,
        getBaseLayers,
        getPanelLayers,
        getMapLayers,
      }}
    >
      {children}
    </MapContext.Provider>
  )
}

const MapPanelContainer = () => (
  <MapContextProvider>
    <MapPanelComponent />
  </MapContextProvider>
)

const MapPanelComponent = () => {
  const {
    baseLayers,
    panelLayers,
    activeBaseLayer,
    activeMapLayers,
    isMapPanelVisible,
    zoomLevel,
    setActiveBaseLayer,
    setActiveMapLayers,
    setVisibleMapLayers,
    setMapPanelVisible,
    getBaseLayers,
    getPanelLayers,
    getMapLayers,
  } = React.useContext(MapContext)

  React.useEffect(() => {
    getBaseLayers()
    getPanelLayers()
    getMapLayers()
  }, [])

  function onLayerToggle(payload) {
    setActiveMapLayers(
      payload.legendItems?.some(({ notSelectable }) => !notSelectable) &&
        payload.legendItems.length > 0
        ? payload.legendItems.map(({ id: legendItemId }) => ({ id: legendItemId, isVisible: true }))
        : [{ id: payload.id, isVisible: true }],
    )
  }

  return (
    <MapPage
      mapBaseLayers={baseLayers}
      panelLayers={panelLayers}
      activeBaseLayer={activeBaseLayer}
      overlays={activeMapLayers}
      onBaseLayerToggle={setActiveBaseLayer}
      onLayerToggle={onLayerToggle}
      onLayerVisibilityToggle={setVisibleMapLayers}
      onMapPanelToggle={setMapPanelVisible}
      isMapPanelVisible={isMapPanelVisible}
      zoomLevel={zoomLevel}
    />
  )
}

export default MapPanelContainer
