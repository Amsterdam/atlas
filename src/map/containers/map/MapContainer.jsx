import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { LatLngBounds } from 'leaflet';

import { isEmbedded, isPrintOrEmbedMode } from '../../../shared/ducks/ui/ui';

import DrawTool from '../../containers/draw-tool/DrawToolContainer';
import ToggleFullscreen from '../../../app/components/ToggleFullscreen/ToggleFullscreen';
import ContextMenu from '../../../app/components/ContextMenu/ContextMenu';

import LeafletContainer from '../leaflet/LeafletContainer';
import MapPanelContainer from '../../containers/panel/MapPanelContainer';
import MapPreviewPanelContainer from '../../containers/preview-panel/MapPreviewPanelContainer';
import MapEmbedButton from '../../components/map-embed-button/MapEmbedButton';
import { previewDataAvailable as previewDataAvailableSelector } from '../../../shared/ducks/selection/selection';
import { getDrawingMode } from '../../ducks/map/selectors';

export const overrideLeafletGetBounds = (map) => {
  // We override here the getBounds method of Leaflet
  // To ensure the full coverage of the visible area
  // of the NonTiledLayer layer types
  map.getBounds = () => { // eslint-disable-line no-param-reassign
    const bounds = map.getPixelBounds();
    const sw = map.unproject(bounds.getBottomLeft());
    const ne = map.unproject(bounds.getTopRight());

    const latLngBounds = (new LatLngBounds(sw, ne)).pad(0.02);
    return latLngBounds;
  };
};

const mapStateToProps = (state) => ({
  drawMode: getDrawingMode(state),
  embedMode: isEmbedded(state),
  printOrEmbedMode: isPrintOrEmbedMode(state),
  previewDataAvailable: previewDataAvailableSelector(state)
});

class MapContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      leafletInstance: null
    };
    this.setLeafletInstance = this.setLeafletInstance.bind(this);
  }

  setLeafletInstance(leafletInstance) {
    overrideLeafletGetBounds(leafletInstance);
    this.setState({ leafletInstance });
  }

  render() {
    const {
      embedMode,
      printOrEmbedMode,
      isFullscreen,
      toggleFullscreen,
      drawMode,
      showPreviewPanel,
      previewDataAvailable
    } = this.props;
    return (
      <div className="qa-map">
        {/* Note: map must not be unmounted when showing the iframe */}
        {/* as it will reset the map's state */}
        <div className={`c-map c-map--drawing-mode-${drawMode} qa-map-container`}>
          <LeafletContainer
            getLeafletInstance={this.setLeafletInstance}
          />
          {
            this.state.leafletInstance && (
              <DrawTool
                leafletInstance={this.state.leafletInstance}
              />
            )
          }
          {toggleFullscreen && (
            <ToggleFullscreen
              isFullscreen={isFullscreen}
              title="Kaart"
              onToggleFullscreen={toggleFullscreen}
            />
          )}
          <MapPanelContainer isMapPanelVisible />
          {
            embedMode ? (
              <MapEmbedButton />
            ) : ''
          }
          { (!printOrEmbedMode && isFullscreen) && <ContextMenu isMapPanelVisible /> }
          {showPreviewPanel && previewDataAvailable && <MapPreviewPanelContainer />}
        </div>
      </div>
    );
  }
}

MapContainer.defaultProps = {
  leafletInstance: null,
  showPreviewPanel: false,
  drawMode: 'none',
  toggleFullscreen: null,
  isFullscreen: true,
  printOrEmbedMode: false
};

MapContainer.propTypes = {
  isFullscreen: PropTypes.bool,
  toggleFullscreen: PropTypes.func,
  drawMode: PropTypes.string,
  embedMode: PropTypes.bool.isRequired,
  printOrEmbedMode: PropTypes.bool,
  showPreviewPanel: PropTypes.bool,
  previewDataAvailable: PropTypes.bool.isRequired
};

export default connect(mapStateToProps)(MapContainer);
