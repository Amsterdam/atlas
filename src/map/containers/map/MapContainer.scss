@import "../../../shared/styles/config";

.bottom-left {
  bottom: $container-whitespace;
  display: flex;
  left: $container-whitespace;
  position: absolute;
  z-index: 9;

  section + section {
    margin-left: $container-whitespace;
  }
}
