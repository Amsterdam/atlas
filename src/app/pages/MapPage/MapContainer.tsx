import { Alert, Heading, Link } from '@amsterdam/asc-ui'
// eslint-disable-next-line import/no-extraneous-dependencies
import { Feature } from 'geojson'
import { FunctionComponent, useEffect, useMemo, useReducer, useState } from 'react'
import { useSelector } from 'react-redux'
import { Link as RouterLink } from 'react-router-dom'
import {
  getMapLayers as fetchMapLayers,
  getPanelLayers as fetchPanelLayers,
  MapCollection,
  MapLayer,
} from '../../../map/services'
import { getUser } from '../../../shared/ducks/user/user'
import { toMap } from '../../../store/redux-first-router/actions'
import useParam from '../../utils/useParam'
import MapContext, { initialState, MapState } from './MapContext'
import MapPage from './MapPage'
import { mapLayersParam, panoFullScreenParam, polygonParam, polylineParam } from './query-params'
import buildLeafletLayers from './utils/buildLeafletLayers'

type Action =
  | { type: 'setPanelLayers'; payload: MapCollection[] }
  | { type: 'setMapLayers'; payload: MapLayer[] }
  | { type: 'setDetailFeature'; payload: Feature | null }
  | { type: 'setPanoImageDate'; payload: string | null }

const reducer = (state: MapState, action: Action): MapState => {
  switch (action.type) {
    case 'setPanelLayers':
      return {
        ...state,
        panelLayers: action.payload,
      }
    case 'setMapLayers':
      return {
        ...state,
        mapLayers: action.payload,
      }
    case 'setDetailFeature':
      return {
        ...state,
        detailFeature: action.payload,
      }
    case 'setPanoImageDate':
      return {
        ...state,
        panoImageDate: action.payload,
      }
    default:
      return state
  }
}

const MapContainer: FunctionComponent = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState)
  const [activeMapLayers] = useParam(mapLayersParam)
  const [polyline] = useParam(polylineParam)
  const [polygon] = useParam(polygonParam)
  const showDrawContent = !!(polyline || polygon)
  const [showDrawTool, setShowDrawTool] = useState(showDrawContent)
  const [panoFullScreen, setPanoFullScreen] = useParam(panoFullScreenParam)
  const user = useSelector(getUser)

  const legendLeafletLayers = useMemo(
    () => buildLeafletLayers(activeMapLayers, state.mapLayers, user),
    [activeMapLayers, state.mapLayers, user],
  )

  function setDetailFeature(payload: Feature | null) {
    dispatch({ type: 'setDetailFeature', payload })
  }

  function setPanoImageDate(payload: string | null) {
    dispatch({ type: 'setPanoImageDate', payload })
  }

  async function getPanelLayers() {
    const panelLayers = await fetchPanelLayers()

    dispatch({
      type: 'setPanelLayers',
      payload: panelLayers,
    })
  }

  async function getMapLayers() {
    const mapLayers = await fetchMapLayers()
    dispatch({
      type: 'setMapLayers',
      payload: mapLayers,
    })
  }

  useEffect(() => {
    Promise.all([getPanelLayers(), getMapLayers()])
      .then(() => {})
      .catch((error: string) => {
        // eslint-disable-next-line no-console
        console.error(`MapContainer: problem fetching panel and map layers: ${error}`)
      })
  }, [])

  return (
    <MapContext.Provider
      value={{
        ...state,
        legendLeafletLayers,
        setDetailFeature,
        showDrawTool,
        setShowDrawTool,
        showDrawContent,
        panoFullScreen,
        setPanoFullScreen,
        setPanoImageDate,
      }}
    >
      <Alert level="info" dismissible>
        <Heading as="h3">Let op: Deze nieuwe interactieve kaart is nog in aanbouw.</Heading>
        <Link darkBackground to={toMap()} as={RouterLink} inList>
          Naar de oude kaart
        </Link>
      </Alert>
      <MapPage>{children}</MapPage>
    </MapContext.Provider>
  )
}

export default MapContainer
