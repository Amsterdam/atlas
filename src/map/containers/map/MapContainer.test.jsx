import React from 'react';
import configureMockStore from 'redux-mock-store';
import { shallow } from 'enzyme';
// import { LatLngBounds, Bounds } from 'leaflet';

import MapContainer, { overrideLeafletGetBounds } from './MapContainer';

let initialState;

describe('overrideLeafletGetBounds', () => {
  it('should overide the getBounds', () => {
    const bl = { x: 482988, y: -551411 };
    const tr = { x: 484747, y: -552666 };
    const pixelBounds = {
      max: { x: 484747, y: -551411 },
      min: { x: 482988, y: -552666 },
      getBottomLeft: () => bl,
      getTopRight: () => tr
    };

    const sw = { lat: 52.35884827870951, lng: 4.877935872162896 };
    const ne = { lat: 52.36841402615222, lng: 4.8995211251421305 };

    const resultBounds = {
      _northEast: {
        lat: 52.369370600896495,
        lng: 4.9016796504400535
      },

      _southWest: {
        lat: 52.35789170396524,
        lng: 4.875777346864972
      }
    };
    const map = {
      getPixelBounds: jest.fn(() => pixelBounds),
      unproject: jest.fn((p) => {
        if (p.x === 482988) return sw;
        return ne;
      })
    };

    overrideLeafletGetBounds(map);
    expect(map.getBounds).toBeDefined();
    const bounds = map.getBounds();
    expect(bounds).toEqual(resultBounds);
  });
});

describe('MapContainer', () => {
  beforeEach(() => {
    initialState = {
      ui: {
        isMapFullscreen: false
      },
      map: {
        drawingMode: 'none'
      }
    };
  });

  it('should render', () => {
    const store = configureMockStore()({ ...initialState });
    const wrapper = shallow(<MapContainer />, { context: { store } }).dive();

    expect(wrapper).toMatchSnapshot();
  });

  it('should render with drawingmode: draw', () => {
    const store = configureMockStore()({ ...initialState, map: { drawingMode: 'draw' } });
    const wrapper = shallow(<MapContainer />, { context: { store } }).dive();

    expect(wrapper).toMatchSnapshot();
  });

  it('should set the leaflet instance state', () => {
    const store = configureMockStore()({ ...initialState });
    const wrapper = shallow(<MapContainer />, { context: { store } }).dive();
    wrapper.instance().setLeafletInstance({});
    expect(wrapper.instance().state.leafletInstance).toBeTruthy();
    expect(wrapper.instance().state.leafletInstance.getBounds).toBeDefined();
  });
});
