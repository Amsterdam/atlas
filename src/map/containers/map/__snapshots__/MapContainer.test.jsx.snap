// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapContainer should render 1`] = `
<div
  className="c-map-page"
  style={
    Object {
      "height": "100%",
    }
  }
>
  <div
    className="qa-map"
    style={Object {}}
  >
    <div
      className="c-map c-map--drawing-mode-none qa-map-container"
    >
      <Connect(LeafletContainer)
        getLeafletInstance={[Function]}
      />
      <ToggleFullscreen
        isFullscreen={false}
        onToggleFullscreen={[Function]}
      />
      <Connect(MapPanelContainer)
        isMapPanelVisible={true}
      />
    </div>
  </div>
</div>
`;

exports[`MapContainer should render with drawingmode: draw 1`] = `
<div
  className="c-map-page"
  style={
    Object {
      "height": "100%",
    }
  }
>
  <div
    className="qa-map"
    style={Object {}}
  >
    <div
      className="c-map c-map--drawing-mode-draw qa-map-container"
    >
      <Connect(LeafletContainer)
        getLeafletInstance={[Function]}
      />
      <ToggleFullscreen
        isFullscreen={false}
        onToggleFullscreen={[Function]}
      />
      <Connect(MapPanelContainer)
        isMapPanelVisible={true}
      />
    </div>
  </div>
</div>
`;
